File: api/src/lib/audit.ts
import { db } from '../db/pool';

interface AuditEventInput {
    actorId?: string | null;
    jobId?: string | null;
    eventType: string;
    oldState?: string;
    newState?: string;
    surface?: 'IDE' | 'WEB' | 'API' | 'SYSTEM';
    metadata?: Record<string, unknown>;
    ipAddress?: string;
}

/**
 * Writes an audit_events row for every state-changing operation.
 * Required per AGENT_DEVELOPMENT_GUIDE.md Section 6.2 Rule 6:
 * "All state-changing operations on refactor_jobs, spec_revisions, patch_attempts
 *  must produce a row in audit_events. Never skip audit logging for performance."
 */
export async function writeAuditEvent(event: AuditEventInput): Promise<void> {
    await db.query(
        `INSERT INTO audit_events
       (actor_id, job_id, event_type, old_state, new_state, surface, metadata, ip_address)
     VALUES ($1, $2, $3, $4, $5, $6, $7, $8::inet)`,
        [
            event.actorId ?? null,
            event.jobId ?? null,
            event.eventType,
            event.oldState ?? null,
            event.newState ?? null,
            event.surface ?? 'API',
            JSON.stringify(event.metadata ?? {}),
            event.ipAddress ?? null,
        ]
    );
}

File: worker/activities/audit.py
"""Audit event activity — writes audit_events table from Temporal activities."""
import logging
from dataclasses import dataclass, field
from typing import Optional
from temporalio import activity
import asyncpg

from config import settings

logger = logging.getLogger(__name__)


@dataclass
class AuditInput:
    job_id: str
    event_type: str
    old_state: Optional[str] = None
    new_state: Optional[str] = None
    metadata: dict = field(default_factory=dict)
    surface: str = "SYSTEM"


@activity.defn
async def write_audit_event(inp: AuditInput) -> None:
    """
    Inserts an audit_events row.
    Also updates refactor_jobs.status to the new state (status = Temporal read projection).
    Per AGENT_DEVELOPMENT_GUIDE.md Section 3.3: status is updated via Temporal activities only.
    """
    activity.heartbeat()

    conn = await asyncpg.connect(settings.database_url)
    try:
        await conn.execute(
            """
            INSERT INTO audit_events (job_id, event_type, old_state, new_state, surface, metadata)
            VALUES ($1,$2,$3,$4,$5,$6::jsonb)
            """,
            inp.job_id,
            inp.event_type,
            inp.old_state,
            inp.new_state,
            inp.surface,
            __import__("json").dumps(inp.metadata),
        )

        # Update refactor_jobs.status (read projection of Temporal state)
        if inp.event_type == "job.state_change" and inp.new_state:
            extra_fields = ""
            if inp.new_state in ("DELIVERED", "FAILED", "FALLBACK_REQUIRED"):
                extra_fields = ", completed_at=NOW()"

            failure_reason = inp.metadata.get("reason") if inp.new_state == "FAILED" else None

            if failure_reason:
                await conn.execute(
                    f"""UPDATE refactor_jobs SET status=$1, failure_reason=$2, updated_at=NOW(){extra_fields}
                        WHERE id=$3""",
                    inp.new_state, failure_reason, inp.job_id,
                )
            else:
                await conn.execute(
                    f"UPDATE refactor_jobs SET status=$1, updated_at=NOW(){extra_fields} WHERE id=$2",
                    inp.new_state, inp.job_id,
                )
    finally:
        await conn.close()
